File: packages/catalog_core/lib/src/catalogs.dart
import 'dart:io';
import 'dart:math';

import 'package:sqlite3/sqlite3.dart';

import 'store.dart';

/// One catalog on disk: a folder holding its database and its photos.
///
/// A catalog is self-contained — its own entries, photos, device id and
/// sync partners — so copying the folder copies the whole catalog. Its
/// display name lives inside it; the registry only caches the name for
/// listing.
class CatalogInfo {
  final String id;
  final String name;
  final Directory dir;

  const CatalogInfo({required this.id, required this.name, required this.dir});

  String get dbPath => '${dir.path}/catlog.db';

  /// Database plus photos, in bytes. Zero for a folder that is gone.
  int get sizeInBytes {
    if (!dir.existsSync()) return 0;
    var total = 0;
    for (final f in dir.listSync(recursive: true)) {
      if (f is File) {
        try {
          total += f.lengthSync();
        } catch (_) {}
      }
    }
    return total;
  }
}

/// Thrown when a name would collide with a catalog that already exists.
class DuplicateCatalogName implements Exception {
  final String name;
  const DuplicateCatalogName(this.name);

  @override
  String toString() => 'A catalog named "$name" already exists';
}

/// Thrown when the one-time move into the multi-catalog layout fails.
/// The old catalog is left where it was and stays openable.
class MigrationFailed implements Exception {
  final String reason;
  const MigrationFailed(this.reason);

  @override
  String toString() =>
      'Could not move the catalog into the new layout: $reason. '
      'Nothing was lost — the old catalog is untouched. Free some space, '
      'check that the app may write to its own folder, and start the app '
      'again.';
}

/// Every catalog on this device, and the settings that belong to the app
/// rather than to any one of them.
///
/// Layout under the root directory:
///
///     app.db                     registry + shared settings
///     catalogs/<id>/catlog.db    one catalog
///     catalogs/<id>/images/      its photos
class CatalogManager implements SharedSettings {
  final Directory root;
  final Database _db;

  CatalogManager._(this.root, this._db);

  /// The language chosen on this device, from the shared settings or —
  /// before the one-time move — from the single catalog that used to
  /// hold it. Read without opening a manager, because the name of the
  /// catalog the move creates has to be in that language.
  static String? savedLocale(String rootPath) {
    for (final (path, key) in [
      ('$rootPath/app.db', 'locale'),
      ('$rootPath/catlog.db', 'u:locale'),
    ]) {
      if (!File(path).existsSync()) continue;
      final db = sqlite3.open(path);
      try {
        final table = key == 'locale' ? 'settings' : 'local_settings';
        final rows =
            db.select('SELECT value FROM $table WHERE key = ?', [key]);
        if (rows.isNotEmpty) return rows.first['value'] as String;
      } catch (_) {
        // A database without the table is simply one that has no answer.
      } finally {
        db.dispose();
      }
    }
    return null;
  }

  /// Opens the manager, creating the layout if needed. An older
  /// single-catalog installation is moved in on the way, named
  /// [defaultName] — the word that used to be the home screen's title.
  factory CatalogManager.open(String rootPath,
      {String defaultName = 'Catalog'}) {
    final root = Directory(rootPath)..createSync(recursive: true);
    final db = sqlite3.open('${root.path}/app.db');
    db.execute('''
      PRAGMA journal_mode = WAL;
      CREATE TABLE IF NOT EXISTS catalogs (
        id      TEXT PRIMARY KEY,
        name    TEXT NOT NULL,
        created TEXT NOT NULL
      );
      CREATE TABLE IF NOT EXISTS settings (
        key   TEXT PRIMARY KEY,
        value TEXT NOT NULL
      );
    ''');
    final manager = CatalogManager._(root, db);
    manager._adopt(defaultName);
    return manager;
  }

  void close() => _db.dispose();

  Directory get _catalogsDir =>
      Directory('${root.path}/catalogs')..createSync(recursive: true);

  // ------------------------------------------------------------- registry

  List<CatalogInfo> catalogs() => [
        for (final r in _db.select(
            'SELECT id, name FROM catalogs ORDER BY created, name'))
          CatalogInfo(
            id: r['id'] as String,
            name: r['name'] as String,
            dir: Directory('${_catalogsDir.path}/${r['id']}'),
          )
      ];

  CatalogInfo? byId(String id) =>
      catalogs().where((c) => c.id == id).firstOrNull;

  /// The catalog to open on launch: the one last used, or the first.
  CatalogInfo get active {
    final all = catalogs();
    final wanted = get('activeCatalog');
    return all.where((c) => c.id == wanted).firstOrNull ?? all.first;
  }

  set active(CatalogInfo catalog) => set('activeCatalog', catalog.id);

  /// Opens a catalog's store, wired to the shared settings.
  CatalogStore openStore(CatalogInfo catalog) {
    catalog.dir.createSync(recursive: true);
    return CatalogStore.open(catalog.dbPath)..shared = this;
  }

  /// Creates an empty catalog. Names are unique so the switcher and the
  /// backup files stay unambiguous.
  CatalogInfo create(String name) {
    final clean = name.trim();
    if (clean.isEmpty) throw ArgumentError('A catalog needs a name');
    _requireFreeName(clean);
    final id = _newId();
    _db.execute(
      'INSERT INTO catalogs (id, name, created) VALUES (?, ?, ?)',
      [id, clean, DateTime.now().toUtc().toIso8601String()],
    );
    final info = CatalogInfo(
        id: id,
        name: clean,
        dir: Directory('${_catalogsDir.path}/$id')..createSync(recursive: true));
    // The name lives in the catalog too, so a copied folder is complete.
    final store = openStore(info);
    store.setLocalSetting(catalogNameKey, clean);
    store.close();
    return info;
  }

  void rename(String id, String name) {
    final clean = name.trim();
    if (clean.isEmpty) throw ArgumentError('A catalog needs a name');
    final current = byId(id);
    if (current == null) return;
    if (current.name != clean) _requireFreeName(clean);
    _db.execute('UPDATE catalogs SET name = ? WHERE id = ?', [clean, id]);
    final store = openStore(current);
    store.setLocalSetting(catalogNameKey, clean);
    store.close();
  }

  /// Removes a catalog and everything in it. The caller writes the
  /// keepsake bundle first — this is the point of no return.
  ///
  /// The catalog currently open is refused: its database is held open by
  /// the app, and deleting the folder underneath it would leave every
  /// cat added afterwards writing into a file that no longer exists.
  /// Switch to another catalog first.
  void delete(String id) {
    final info = byId(id);
    if (info == null) return;
    if (catalogs().length == 1) {
      throw StateError('The last catalog cannot be deleted');
    }
    if (active.id == id) {
      throw StateError('Switch to another catalog before deleting this one');
    }
    // Files first: if they cannot go, the catalog stays listed and the
    // keeper is told, rather than a folder nobody can reach any more.
    if (info.dir.existsSync()) info.dir.deleteSync(recursive: true);
    _db.execute('DELETE FROM catalogs WHERE id = ?', [id]);
  }

  void _requireFreeName(String name) {
    final taken = catalogs()
        .any((c) => c.name.toLowerCase() == name.toLowerCase());
    if (taken) throw DuplicateCatalogName(name);
  }

  // ------------------------------------------------------ shared settings

  @override
  String? get(String key) {
    final rows =
        _db.select('SELECT value FROM settings WHERE key = ?', [key]);
    return rows.isEmpty ? null : rows.first['value'] as String;
  }

  @override
  void set(String key, String value) => _db.execute(
        'INSERT INTO settings (key, value) VALUES (?, ?) '
        'ON CONFLICT(key) DO UPDATE SET value = excluded.value',
        [key, value],
      );

  @override
  void remove(String key) =>
      _db.execute('DELETE FROM settings WHERE key = ?', [key]);

  @override
  List<(String, String)> byPrefix(String prefix) => [
        for (final r in _db.select(
            'SELECT key, value FROM settings WHERE key LIKE ?',
            ['$prefix%']))
          (
            (r['key'] as String).substring(prefix.length),
            r['value'] as String
          )
      ];

  // ------------------------------------------------------------ migration

  /// Moves a pre-catalogs installation — `catlog.db` and `images/` in
  /// the root — into the new layout, and copies the settings that are
  /// now shared out of it. Does nothing once a catalog is registered.
  void _adopt(String defaultName) {
    if (catalogs().isNotEmpty) return;
    final oldDb = File('${root.path}/catlog.db');
    if (!oldDb.existsSync()) {
      create(defaultName);
      return;
    }
    _checkpoint(oldDb);
    final id = _newId();
    final dir = Directory('${_catalogsDir.path}/$id');
    var committed = false;
    try {
      dir.createSync(recursive: true);
      // Copy first, verify, and only then remove the original: an
      // interrupted move must leave a catalog that still opens.
      oldDb.copySync('${dir.path}/catlog.db');
      final oldImages = Directory('${root.path}/images');
      if (oldImages.existsSync()) {
        _copyDir(oldImages, Directory('${dir.path}/images'));
      }
      final moved = CatalogStore.open('${dir.path}/catlog.db');
      final name = moved.localSetting(catalogNameKey) ?? defaultName;
      moved.setLocalSetting(catalogNameKey, name);
      _liftSharedSettings(moved);
      moved.close();
      _db.execute(
        'INSERT INTO catalogs (id, name, created) VALUES (?, ?, ?)',
        [id, name, DateTime.now().toUtc().toIso8601String()],
      );
      // Past this line the copy is the catalog: tidying up the old
      // location must never be able to take it away again, so failures
      // from here on leave litter rather than a hole.
      committed = true;
      oldDb.deleteSync();
      for (final side in ['-wal', '-shm']) {
        try {
          final f = File('${oldDb.path}$side');
          if (f.existsSync()) f.deleteSync();
        } catch (_) {}
      }
      try {
        if (oldImages.existsSync()) oldImages.deleteSync(recursive: true);
      } catch (_) {}
    } catch (e) {
      if (committed) return;
      // Leave nothing half-moved: the original stays where it was and
      // the next launch tries again.
      try {
        if (dir.existsSync()) dir.deleteSync(recursive: true);
      } catch (_) {}
      _db.execute('DELETE FROM catalogs WHERE id = ?', [id]);
      throw MigrationFailed('$e');
    }
  }

  /// Folds the write-ahead log back into the database so the side files
  /// carry nothing that would be lost by moving the database alone.
  void _checkpoint(File db) {
    final handle = sqlite3.open(db.path);
    try {
      handle.execute('PRAGMA wal_checkpoint(TRUNCATE)');
    } finally {
      handle.dispose();
    }
  }

  /// Copies the settings that belong to the app out of the catalog that
  /// used to be the only one, so nothing is asked for twice.
  void _liftSharedSettings(CatalogStore from) {
    final author = from.author;
    if (author != null && author.isNotEmpty && get('author') == null) {
      set('author', author);
    }
    for (final (key, value) in from.allLocalSettings()) {
      if (isSharedSetting(key) && get(key) == null) set(key, value);
    }
  }

  void _copyDir(Directory from, Directory to) {
    to.createSync(recursive: true);
    for (final f in from.listSync()) {
      if (f is File) f.copySync('${to.path}/${f.uri.pathSegments.last}');
    }
  }

  static final _random = Random.secure();

  String _newId() => List.generate(
      16, (_) => _random.nextInt(256).toRadixString(16).padLeft(2, '0')).join();
}

/// Where a catalog keeps its own display name.
const catalogNameKey = 'catalogName';

File: desktop/crates/catlog-core/src/catalogs.rs
//! Every Catalog on this device. One Catalog is a folder holding its
//! database and its photos, self-contained: copying the folder copies
//! the whole Catalog. Its display name lives inside it; the registry
//! only lists the names and remembers which one was open last.
//!
//! Layout under the root directory:
//!
//! ```text
//! registry.json                the list and the active one
//! catalogs/<id>/catalog.db     one Catalog
//! catalogs/<id>/images/        its photos
//! ```

use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};

use crate::Result;
use crate::catalog::Catalog;
use crate::error::Error;

/// Where a Catalog keeps its own display name.
pub const CATALOG_NAME_KEY: &str = "catalogName";

/// One Catalog on disk.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct CatalogInfo {
    pub id: String,
    pub name: String,
    pub created: String,
}

#[derive(Debug, Default, Serialize, Deserialize)]
struct Registry {
    catalogs: Vec<CatalogInfo>,
    active: Option<String>,
}

pub struct CatalogManager {
    root: PathBuf,
    registry: Registry,
}

impl CatalogManager {
    /// Opens the manager, creating the layout and a first Catalog named
    /// `default_name` when there is none.
    pub fn open(root: &Path, default_name: &str) -> Result<CatalogManager> {
        std::fs::create_dir_all(root.join("catalogs")).map_err(|e| Error::io(root, e))?;
        let path = root.join("registry.json");
        let registry = match std::fs::read_to_string(&path) {
            Ok(text) => serde_json::from_str(&text)?,
            Err(_) => Registry::default(),
        };
        let mut manager = CatalogManager {
            root: root.to_path_buf(),
            registry,
        };
        if manager.registry.catalogs.is_empty() {
            manager.create(default_name)?;
        }
        Ok(manager)
    }

    fn save(&self) -> Result<()> {
        let path = self.root.join("registry.json");
        let text = serde_json::to_string_pretty(&self.registry)?;
        std::fs::write(&path, text).map_err(|e| Error::io(&path, e))
    }

    pub fn root(&self) -> &Path {
        &self.root
    }

    /// The Catalog's folder.
    pub fn dir(&self, id: &str) -> PathBuf {
        self.root.join("catalogs").join(id)
    }

    /// Every Catalog, oldest first.
    pub fn catalogs(&self) -> &[CatalogInfo] {
        &self.registry.catalogs
    }

    pub fn by_id(&self, id: &str) -> Option<&CatalogInfo> {
        self.registry.catalogs.iter().find(|c| c.id == id)
    }

    /// The Catalog to open on launch: the one last used, or the first.
    pub fn active(&self) -> &CatalogInfo {
        self.registry
            .active
            .as_deref()
            .and_then(|id| self.by_id(id))
            .unwrap_or(&self.registry.catalogs[0])
    }

    pub fn set_active(&mut self, id: &str) -> Result<()> {
        if self.by_id(id).is_none() {
            return Err(Error::Invalid(format!("No catalog {id}")));
        }
        self.registry.active = Some(id.to_string());
        self.save()
    }

    /// Opens a Catalog's store.
    pub fn open_store(&self, info: &CatalogInfo) -> Result<Catalog> {
        Catalog::open(&self.dir(&info.id))
    }

    /// Creates an empty Catalog. Names are unique so the switcher and
    /// the backup files stay unambiguous.
    pub fn create(&mut self, name: &str) -> Result<CatalogInfo> {
        let clean = name.trim();
        if clean.is_empty() {
            return Err(Error::Invalid("A catalog needs a name".into()));
        }
        self.require_free_name(clean)?;
        let mut bytes = [0u8; 16];
        getrandom::fill(&mut bytes)
            .map_err(|e| Error::Invalid(format!("no randomness for a catalog id: {e}")))?;
        let info = CatalogInfo {
            id: hex::encode(bytes),
            name: clean.to_string(),
            created: chrono::Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Micros, true),
        };
        // The name lives in the Catalog too, so a copied folder is complete.
        let store = self.open_store(&info)?;
        store.set_local_setting(CATALOG_NAME_KEY, clean)?;
        drop(store);
        self.registry.catalogs.push(info.clone());
        if self.registry.active.is_none() {
            self.registry.active = Some(info.id.clone());
        }
        self.save()?;
        Ok(info)
    }

    pub fn rename(&mut self, id: &str, name: &str) -> Result<()> {
        let clean = name.trim();
        if clean.is_empty() {
            return Err(Error::Invalid("A catalog needs a name".into()));
        }
        let Some(current) = self.by_id(id).cloned() else {
            return Ok(());
        };
        if current.name != clean {
            self.require_free_name(clean)?;
        }
        let store = self.open_store(&current)?;
        store.set_local_setting(CATALOG_NAME_KEY, clean)?;
        drop(store);
        if let Some(c) = self.registry.catalogs.iter_mut().find(|c| c.id == id) {
            c.name = clean.to_string();
        }
        self.save()
    }

    /// Removes a Catalog and everything in it. The caller writes the
    /// keepsake bundle first: this is the point of no return. The last
    /// Catalog and the active one are refused.
    pub fn delete(&mut self, id: &str) -> Result<()> {
        let Some(info) = self.by_id(id).cloned() else {
            return Ok(());
        };
        if self.registry.catalogs.len() == 1 {
            return Err(Error::Invalid("The last catalog cannot be deleted".into()));
        }
        if self.active().id == id {
            return Err(Error::Invalid(
                "Switch to another catalog before deleting this one".into(),
            ));
        }
        let dir = self.dir(&info.id);
        if dir.exists() {
            std::fs::remove_dir_all(&dir).map_err(|e| Error::io(&dir, e))?;
        }
        self.registry.catalogs.retain(|c| c.id != id);
        self.save()
    }

    fn require_free_name(&self, name: &str) -> Result<()> {
        let taken = self
            .registry
            .catalogs
            .iter()
            .any(|c| c.name.to_lowercase() == name.to_lowercase());
        if taken {
            return Err(Error::Invalid(format!(
                "A catalog named \"{name}\" already exists"
            )));
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_manager_starts_with_one_catalog_and_manages_more() {
        let dir = tempfile::tempdir().unwrap();
        let mut m = CatalogManager::open(dir.path(), "Clowders").unwrap();
        assert_eq!(m.catalogs().len(), 1);
        assert_eq!(m.active().name, "Clowders");
        let store = m.open_store(m.active()).unwrap();
        assert_eq!(
            store.local_setting(CATALOG_NAME_KEY).as_deref(),
            Some("Clowders")
        );
        drop(store);
        let leipzig = m.create(" Leipzig ").unwrap();
        assert_eq!(leipzig.name, "Leipzig");
        assert!(m.create("leipzig").is_err(), "names are unique, case aside");
        assert!(m.create("  ").is_err());
        assert!(m.dir(&leipzig.id).join("catalog.db").exists());
        m.rename(&leipzig.id, "Leipzig Nord").unwrap();
        m.rename(&leipzig.id, "Leipzig Nord").unwrap();
        assert!(m.rename(&leipzig.id, "Clowders").is_err());
        assert!(m.rename(&leipzig.id, "").is_err());
        m.rename("nobody", "x").unwrap();
        assert_eq!(m.by_id(&leipzig.id).unwrap().name, "Leipzig Nord");
        // The registry survives reopening; the active catalog is remembered.
        m.set_active(&leipzig.id).unwrap();
        assert!(m.set_active("nobody").is_err());
        let m2 = CatalogManager::open(dir.path(), "ignored").unwrap();
        assert_eq!(m2.catalogs().len(), 2);
        assert_eq!(m2.active().id, leipzig.id);
        assert!(m2.root().join("registry.json").exists());
        // Deleting: not the active one, not the last one.
        let mut m2 = m2;
        assert!(m2.delete(&leipzig.id).is_err());
        let first = m2.catalogs()[0].id.clone();
        m2.delete(&first).unwrap();
        assert_eq!(m2.catalogs().len(), 1);
        assert!(!m2.dir(&first).exists());
        assert!(m2.delete(&leipzig.id).is_err(), "the last one stays");
        m2.delete("nobody").unwrap();
    }

    #[test]
    fn a_broken_registry_is_an_error_not_a_crash() {
        let dir = tempfile::tempdir().unwrap();
        std::fs::write(dir.path().join("registry.json"), "junk").unwrap();
        assert!(CatalogManager::open(dir.path(), "x").is_err());
    }
}
